packages/compass-components/src/components/virtual-grid.tsx

```diff
@@ -0,0 +1,287 @@
+/* eslint-disable react/prop-types */
+import React, {
+  createContext,
+  forwardRef,
+  useCallback,
+  useContext,
+  useMemo,
+  useRef,
+} from 'react';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { FixedSizeList } from 'react-window';
+import { useDOMRect } from '../hooks/use-dom-rect';
+import {
+  useVirtualGridArrowNavigation,
+  useVirtualRovingTabIndex,
+} from '../hooks/use-virtual-grid';
+import { mergeProps } from '../utils/merge-props';
+
+type RenderItem = React.FunctionComponent<
+  React.HTMLProps<HTMLDivElement> & {
+    index: number;
+    ['data-vlist-item-idx']: number;
+  }
+>;
+type RenderHeader = React.FunctionComponent;
+
+type VirtualGridProps = {
+  /**
+   * Grid item minimum width (items containers will autostretch to the grid
+   * container size)
+   */
+  itemMinWidth: number;
+  /**
+   * Grid row height
+   */
+  itemHeight: number;
+  /**
+   * Number of items in the grid
+   */
+  itemsCount: number;
+  /**
+   * Number of columns (default is fitting as much as possible on the screen
+   * based on itemMinWidth)
+   */
+  colCount?: number;
+  /**
+   * Render method for grid cell content. Receives props that should be passed
+   * directly to the cell root node for the keyboard navigation to work
+   * correctly
+   */
+  renderItem: RenderItem;
+  /**
+   * Header content height
+   */
+  headerHeight?: number;
+  /**
+   * Render method for grid header component. This should be used if header
+   * content should be scrollable as part of the whole grid
+   */
+  renderHeader?: RenderHeader;
+  /**
+   * Number of rows rendered off-screen (default: 3)
+   */
+  overscanCount?: number;
+  /**
+   * Custom styling for individual elements within the grid DOM.
+   */
+  classNames?: {
+    container?: string;
+    header?: string;
+    grid?: string;
+    row?: string;
+    cell?: string;
+  };
+};
+
+const GridContext = createContext<
+  Pick<
+    Required<VirtualGridProps>,
+    'headerHeight' | 'renderHeader' | 'renderItem' | 'itemsCount'
+  > &
+    Pick<VirtualGridProps, 'classNames'> & {
+      rowCount: number;
+      colCount: number;
+      currentTabbable: number;
+      gridProps: React.HTMLProps<HTMLDivElement>;
+    }
+>({
+  headerHeight: 0,
+  renderHeader: () => null,
+  renderItem: () => null,
+  itemsCount: 0,
+  rowCount: 0,
+  colCount: 0,
+  currentTabbable: 0,
+  gridProps: {},
+});
+
+const GridWithHeader = forwardRef<
+  HTMLDivElement,
+  React.HTMLProps<HTMLDivElement>
+>(function GridHeader({ style, children, ...props }, ref) {
+  const { headerHeight, renderHeader, gridProps, classNames } =
+    useContext(GridContext);
+  return (
+    <div
+      ref={ref}
+      style={{
+        ...style,
+        height: parseFloat(style?.height as string) + headerHeight,
```

Comment:
Can height be undefined? Should we check for NaN?

Reply:
Thanks for highlighting this one. It can't really in this case, this was an example in the react-window library that I copy-pasted, I'll type this a bit better

Reply:
Done in b1258fcfc5bd421019febd23bcbc26915a6c52ee